bot/exts/info/doc/_batch_parser.py

```diff
@@ -0,0 +1,206 @@
+
+        self.stale_inventory_notifier = StaleInventoryNotifier()
+
+    async def get_markdown(self, doc_item: _cog.DocItem) -> str:
+        """
+        Get the result Markdown of `doc_item`.
+
+        If no symbols were fetched from `doc_item`s page before,
+        the HTML has to be fetched and then all items from the page are put into the parse queue.
+
+        Not safe to run while `self.clear` is running.
+        """
+        if doc_item not in self._item_futures:
+            self._item_futures.update((item, ParseResultFuture()) for item in self._page_doc_items[doc_item.url])
+            self._item_futures[doc_item].user_requested = True
+
+            async with bot.instance.http_session.get(doc_item.url) as response:
+                soup = await bot.instance.loop.run_in_executor(
+                    None,
+                    BeautifulSoup,
+                    await response.text(encoding="utf8"),
+                    "lxml",
+                )
+
+            self._queue[:0] = (QueueItem(item, soup) for item in self._page_doc_items[doc_item.url])
+            log.debug(f"Added items from {doc_item.url} to parse queue.")
+
+            if self._parse_task is None:
+                self._parse_task = asyncio.create_task(self._parse_queue())
+        else:
+            self._item_futures[doc_item].user_requested = True
+        with suppress(ValueError):
+            # If the item is not in the list then the item is already parsed or is being parsed
+            self._move_to_front(doc_item)
+        return await self._item_futures[doc_item]
+
+    async def _parse_queue(self) -> None:
+        """
+        Parse all items from the queue, setting their result Markdown on the futures and sending them to redis.
+
+        The coroutine will run as long as the queue is not empty, resetting `self._parse_task` to None when finished.
+        """
+        log.trace("Starting queue parsing.")
+        try:
+            while self._queue:
+                item, soup = self._queue.pop()
+                try:
+                    if (future := self._item_futures[item]).done():
+                        # Some items are present in the inventories multiple times under different symbol names,
+                        # if we already parsed an equal item, we can just skip it.
+                        continue
+
+                    markdown = await bot.instance.loop.run_in_executor(None, get_symbol_markdown, soup, item)
+                    if markdown is not None:
+                        scheduling.create_task(doc_cache.set(item, markdown))
+                    else:
+                        scheduling.create_task(self.stale_inventory_notifier.send_warning(item))
+                except Exception as e:
+                    log.exception(f"Unexpected error when handling {item}")
+                    future.set_exception(e)
+                else:
+                    future.set_result(markdown)
+                await asyncio.sleep(0.1)
```

Comment:
I left it in as without it the bot will consume all available resources for a task that's really low priority - at the cost of maybe delaying the next user request by up to 100 ms; can remove it if that won't be an issue.

Reply:
Maybe that was the case back at 758dd3ef6ca5c1cd7615f0eb6688d7d2f19578ea, but you've since added multiple io-bound coroutines. I doubt it's still an issue, but I have not tried it.

Reply:
We use the tasks here that you mentioned in a comment to prevent locking up the queue waiting for one of them, so the only await here is for the cpu bound executor. Locally on a 4 thread CPU the process takes up roughly 33% of the overall usage

Reply:
Okay, never mind.